compiler/rustc_expand/src/mbe/macro_parser.rs

```diff
@@ -254,24 +254,24 @@ crate enum ParseResult<T> {
 /// of metavars to the token trees they bind to.
 crate type NamedParseResult = ParseResult<FxHashMap<MacroRulesNormalizedIdent, NamedMatch>>;
 
-/// Count how many metavars are named in the given matcher `ms`.
-pub(super) fn count_names(ms: &[TokenTree]) -> usize {
-    ms.iter().fold(0, |count, elt| {
-        count
-            + match elt {
-                TokenTree::Delimited(_, delim) => count_names(delim.inner_tts()),
+/// Count how many metavars are named `matcher`.
+pub(super) fn count_metavars(matcher: &[TokenTree]) -> usize {
+    matcher
+        .iter()
+        .map(|tt| {
+            match tt {
+                TokenTree::Delimited(_, delim) => count_metavars(delim.inner_tts()),
                 TokenTree::MetaVar(..) => 0,
                 TokenTree::MetaVarDecl(..) => 1,
-                // Panicking here would abort execution because `parse_tree` makes use of this
-                // function. In other words, RHS meta-variable expressions eventually end-up here.
-                //
-                // `0` is still returned to inform that no meta-variable was found. `Meta-variables
-                // != Meta-variable expressions`
+                // RHS meta-variable expressions eventually end-up here. `0` is returned to inform that
+                // no meta-variable was found, because "meta-variables" != "meta-variable
+                // expressions".
                 TokenTree::MetaVarExpr(..) => 0,
```

Comment:
This should be unreachable `count_metavars` should never be called when parsing RHS (`parsing_patterns == false`) in the first place, it's 0 there by definition.

Reply:
I originally thought that too, but if you make it unreachable these tests fail:
```
failures:
    [ui] ui/macros/rfc-3086-metavar-expr/count-and-length-are-distinct.rs
    [ui] ui/macros/rfc-3086-metavar-expr/feature-gate-macro_metavar_expr.rs
    [ui] ui/macros/rfc-3086-metavar-expr/macro-expansion.rs
    [ui] ui/macros/rfc-3086-metavar-expr/out-of-bounds-arguments.rs
```
There may be something to clean up here, but I'd prefer to do it in a follow-up. This change was just meant to give this function a nicer name and slightly cleaner code, it wasn't doing anything about where/how it is called.

Reply:
Ok.